server/enterprise/message_export/csv_export/csv_export.go

```diff
@@ -31,8 +31,61 @@ const (
 	CSVWarningFilename       = "warning.txt"
 )
 
+type Row struct {
+	CreateAt           int64
+	TeamId             string
+	TeamName           string
+	TeamDisplayName    string
+	ChannelId          string
+	ChannelName        string
+	ChannelDisplayName string
+	ChannelType        string
+	UserId             string
+	UserEmail          string
+	Username           string
+	PostId             string
+	EditedByPostId     string
+	RepliedToPostId    string
+	PostMessage        string
+	PostType           string
+	UserType           string
+	PreviewsPostId     string
+}
+
 func CsvExport(rctx request.CTX, p shared.ExportParams) (shared.RunExportResults, error) {
-	results := shared.RunExportResults{}
+	// Build the channel exports for the channels that had post or user join/leave activity this batch.
+	genericChannelExports, metadata, results, err := shared.GetGenericExportData(p)
+	if err != nil {
+		return results, err
+	}
+
+	totalRows := results.CreatedPosts + results.EditedOrigMsgPosts + results.DeletedPosts + results.EditedNewMsgPosts +
+		results.UpdatedPosts + results.UploadedFiles + results.DeletedFiles
+	rows := make([]Row, 0, totalRows)
+	for _, channel := range genericChannelExports {
+		rows = append(rows, getJoinLeavePosts(channel)...)
```

Comment:
Is there an easy way of allocating join/leave messages like above? This is quite nitpicking but this is still prone to a few more allocations.

Reply:
Yep, we can record the number of joins & leaves and use that. Done! (I only preallocated joins, because leaves also includes closed out leaves, which csv doesn't export).